in-progress/11103-forwarder-contract/design.md

```diff
@@ -0,0 +1,201 @@
+
+In essence, this class is an API for L1 transactions for the prover node, and a simple wrapper around the `L1TxUtils` class.
+
+### SequencerClient `L1TxManager`
+
+The `SequencerClient` will have a `L1TxManager` that has many of the same functions currently within the `l1-publisher.ts`, but will have different semantics.
+
+The `L1TxManager` will have:
+
+- `queuedRequests: L1TxRequest[]`
+- knowledge of the sequencer's forwarder contract
+
+It will have an interface of:
+
+```typescript
+interface L1TxManager {
+  addRequest(request: L1TxRequest): void;
+  sendRequests(): Promise<TransactionReceipt | undefined>;
+}
+```
+
+The `Sequencer` uses its `L1TxManager.addRequest()` to push requests to the `queuedRequests` list whenever it wants to:
+
+- propose an l2 block
+- cast a governance proposal vote
+- cast a slashing vote
+- claim an epoch proof quote
+
+At end of every iteration of the Sequencer's work loop, it will await a call to `L1TxManager.sendRequests()`, which will send the queued requests to the forwarder contract, and flush the `queuedRequests` list.
+
+### Cancellation/Resend
+
+A complication is that ethereum nodes make replacement of blob transactions expensive, and cancelation impossible, as they operate under the assumption that rollups seldom/never need to replace/cancel blob transactions.
+
+See [geth's blob pool](https://github.com/ethereum/go-ethereum/blob/581e2140f22566655aa8fb2d1e9a6c4a740d3be1/core/txpool/blobpool/blobpool.go) for details/constraints.
+
+This is not true for Aztec's decentralized sequencer set with strict L1 timeliness requirements on L2 blocks.
+
+So a concern is the following scenario:
+
+- proposer A submits a tx with nonce 1 (with a blob) that is not priced aggressively enough
+- Tx1 sits in the blob pool, but is not included in an L1 block
+- proposer A tries to submit another transaction, but needs to know to use Tx2
+- Tx1 needs to be replaced with a higher fee, but it will revert if the network is in a different L2 slot and the bundle contained a proposal
+
+This is addressed by:
+
+- Upgrading viem to at least v2.15.0 to use their nonceManager to be aware of pending nonces
+- Aggressive pricing of blob transactions
+- The L1TxUtils will be able to speed up Tx1 (even if it reverts), which should unblock Tx2
+
+### Setup
+
+There will be an optional environment variable `sequencer.custForwarderContractAddress` that can be used to specify a custom forwarder contract address.
+
+If this is not set, the sequencer will deploy the Aztec Labs implementation of the forwarder contract, using the Universal Deterministic Deployer, supplying the sequencer's address as the deployment salt, and the sequencer's address as the owner.
+
+### Gas
+
+Once the L2 block body is removed from calldata, the "static" arguments to call the propose function should be under 1KB.
```

Comment:
I know you say static here, but signatures still are being passed around.
`65 * (128 * 2 / 3) + 128 / 3 * 32` ~7kb. 

Also,